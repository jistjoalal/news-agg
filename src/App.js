import React, { Component } from 'react';
import axios from 'axios';
import { sortBy } from 'lodash';
import classNames from 'classnames';
import './App.css';

const DEFAULT_QUERY = 'redux';
const DEFAULT_HPP = '10';

const PATH_BASE = 'https://hn.algolia.com/api/v1';
const PATH_SEARCH = '/search';
const PARAM_SEARCH = 'query=';
const PARAM_PAGE = 'page=';
const PARAM_HPP = 'hitsPerPage=';

const SORTS = {
  NONE: list => list,
  TITLE: list => sortBy(list, 'title'),
  AUTHOR: list => sortBy(list, 'author'),
  COMMENTS: list => sortBy(list, 'num_comments').reverse(),
  POINTS: list => sortBy(list, 'points').reverse(),
};

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      results: null,
      searchKey: '',
      searchTerm: DEFAULT_QUERY,
      error: null,
      isLoading: false,
      sortKey: 'NONE',
      isSortReverse: false,
    };

    this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this);
    this.setSearchTopStores = this.setSearchTopStores.bind(this);
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
    this.onDismiss = this.onDismiss.bind(this);
    this.onSort = this.onSort.bind(this);
  }
  
  componentDidMount() {
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });
    this.fetchSearchTopStories(searchTerm);
  }
  
  render() {
    const { searchTerm, results, searchKey, error, isLoading,
      sortKey, isSortReverse }  = this.state;
    const page = (results && results[searchKey]
      && results[searchKey].page) || 0;
    const list = (results && results[searchKey]
      && results[searchKey].hits) || [];
    
    return (
      <div className="page">
        <div className="interactions">
          <Search
            value={searchTerm}
            onChange={this.onSearchChange}
            onSubmit={this.onSearchSubmit}
          >
            Search
          </Search>
          <TableWithError
            list={list}
            onDismiss={this.onDismiss}
            sortKey={sortKey}
            onSort={this.onSort}
            isSortReverse={isSortReverse}
            error={error}
          />
        </div>
        <div className="interactions">
          <ButtonWithLoading
            isLoading={isLoading}
            onClick={() => this.fetchSearchTopStories(searchKey, page + 1)}
          >
            More
          </ButtonWithLoading>
        </div>
      </div>
    );
  }

  onSort(sortKey) {
    const { isSortReverse } = this.state;
    // toggle if reverse sort if already clicked
    const newISR = this.state.sortKey === sortKey && !isSortReverse;
    this.setState({ sortKey, isSortReverse: newISR });
  }
  
  // control search input
  onSearchChange = event => {
    this.setState({ searchTerm: event.target.value });
  }

  // submit search input
  onSearchSubmit = e => {
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });

    if (this.needsToSearchTopStories(searchTerm)) {
      this.fetchSearchTopStories(searchTerm);
    }
    e.preventDefault();
  }

  // remove hit from result list
  onDismiss = id => {
    const { searchKey, results } = this.state;
    const { hits, page } = results[searchKey];

    const isNotId = item => item.objectID !== id;
    const updatedHits = hits.filter(isNotId);
    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updatedHits, page }
      }
    });
  }

  // is result for searchTerm cached already?
  needsToSearchTopStories = searchTerm => !this.state.results[searchTerm];

  // makes request to HN API
  fetchSearchTopStories = (searchTerm, page = 0) => {
    this.setState({ isLoading: true });

    const url = `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}
      &${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`
    axios(url)
      .then(result => this.setSearchTopStores(result.data))
      .catch(error => this.setState({ error, isLoading: false }))
  }
  
  // stores response in state.results, appending to existing hits
  setSearchTopStores = response => {
    const { hits, page } = response;
    const { searchKey, results } = this.state;
    const oldHits = results && results[searchKey]
      ? results[searchKey].hits : [];
    const updatedHits = [...oldHits, ...hits];
    this.setState({ 
      results: { 
        ...results,
        [searchKey]: { hits: updatedHits, page }
      },
      isLoading: false
    });
  }
}

// a search bar
class Search extends Component {
  componentDidMount() {
    if (this.input) { this.input.focus(); }
  }
  render() {
    const { value, onChange, onSubmit, children } = this.props;
    return (
      <form onSubmit={onSubmit}>
        <input
          type="text"
          value={value}
          onChange={onChange}
          ref={el => this.input = el}
        />
        <button type="submit">
          {children}
        </button>
      </form>
    );
  }
}

// table formatted for results from HN API
const Table = ({
  list, onDismiss, sortKey = 'NONE', onSort, isSortReverse
}) => {
  const [ lg, md, sm ] = [{width: '40%'}, {width: '30%'}, {width: '10%'}];
  const sortedList = SORTS[sortKey](list);
  const toggleSortedList = isSortReverse ? sortedList.reverse() : sortedList;
  return (
    <div className="table">
      <div className="table-header">
        <span style={lg}>
          <Sort sortKey={'TITLE'} onSort={onSort} activeSortKey={sortKey}
            isSortReverse={isSortReverse}
          >
            Title
          </Sort>
        </span>
        <span style={md}>
          <Sort sortKey={'AUTHOR'} onSort={onSort} activeSortKey={sortKey}
            isSortReverse={isSortReverse}
          >
            Author
          </Sort>
        </span>
        <span style={sm}>
          <Sort sortKey={'COMMENTS'} onSort={onSort} activeSortKey={sortKey}
            isSortReverse={isSortReverse}
          >
            Comments
          </Sort>
        </span>
        <span style={sm}>
          <Sort sortKey={'POINTS'} onSort={onSort} activeSortKey={sortKey}
            isSortReverse={isSortReverse}
          >
            Points
          </Sort>
        </span>
        <span style={sm}>
          Archive
        </span>
      </div>
      {toggleSortedList.map(item => {
        return (
          <div key={item.objectID} className="table-row">
            <span style={lg}>
              <a href={item.url}>{item.title}</a>
            </span>
            <span style={md}>
              {item.author}
            </span>
            <span style={sm}>
              {item.num_comments}
            </span>
            <span style={sm}>
              {item.points}
            </span>
            <span style={sm}>
              <Button
                onClick={() => onDismiss(item.objectID)}
                className="button-inline"
              >
                Dismiss
              </Button>
            </span>
          </div>
        );
      })}
    </div>
  );
}

const Sort = ({ sortKey, onSort, children, activeSortKey, isSortReverse }) => {
  const active = sortKey === activeSortKey;
  const direction = isSortReverse ? 'down' : 'up';
  const sortClass = classNames('button-inline', { 'button-active': active });
  return (
    <span>
      <Button onClick={() => onSort(sortKey)} className={sortClass}>
        {children}
      </Button>
      <ArrowWithNull show={active} direction={direction} />
    </span>
  );
}


const Arrow = ({ direction }) =>
  <i class={`fa fa-arrow-${direction}`}></i>

// a button w/ a click listener
const Button = ({ onClick, className = '', children }) => 
    <button onClick={onClick} className={className}>
      {children}
    </button>

const withLoading = Component => ({ isLoading, ...rest }) =>
  isLoading ?
    <i className="fa fa-spinner fa-spin fa-3x"></i>
  : <Component { ...rest } />

const withError = Component => ({ error, ...rest }) =>
  error ?
    <div className="interactions">
      <p>Something went wrong.</p>
    </div>
  : <Component { ...rest } />

const withNull = Component => ({ show, ...rest }) =>
  show ?
    <Component { ...rest } />
  : null;

const ButtonWithLoading = withLoading(Button);
const TableWithError = withError(Table);
const ArrowWithNull = withNull(Arrow);

export default App;

export { Button, Search, Table, SORTS };