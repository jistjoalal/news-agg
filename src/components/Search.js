import React, { Component } from 'react';

// a search bar
class Search extends Component {

  render() {
    const { value, onChange, onSubmit, onFocus, children } = this.props;
    return (
      <form className="search" onSubmit={onSubmit}>
        <input
          type="text"
          value={value}
          onChange={onChange}
          onFocus={onFocus}
        />
        <button className="button-clickable" type="submit">
          {children}
        </button>
      </form>
    );
  }
}

export default Search;