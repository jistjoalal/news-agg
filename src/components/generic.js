import React from 'react';
import classNames from 'classnames';

// generic.js:
// - components and helpers

/**
 * HOCS
 */

const withError = Component => ({ error, ...rest }) =>
  !error ? <Component { ...rest } />
  : <div className="interactions">
      <p>Something went wrong.</p>
    </div>

const withLoading = Component => ({ isLoading, ...rest }) =>
  !isLoading ? <Component { ...rest } />
  : <i className="fa fa-spinner fa-spin fa-3x"></i>

/**
 * Font Awesome
 */

const ICON_FA = {
  'HN': 'hacker-news',
  'Reddit': 'reddit-alien'
}

const ICONS = ({ source }) =>
  // default to text
  !ICON_FA[source] ? source
  : <i className={`fa fa-${ICON_FA[source]}`}></i>

const SourceIcon = ({ source }) => 
  <ICONS source={source} />

/**
 * clickables
 */

// direction: true = up
const SortArrow = ({ show, direction }) => {
  const isUp = show && direction;
  const isDown = show && !direction;
  return (
    <span className="SortArrow">
      <Arrow show={isUp} direction={true} />
      <Arrow show={isDown} direction={false} />
    </span>
  )
}

const Arrow = ({ show, direction }) => {
  const className = classNames('fa', 
    {'fa-sort-desc': direction},
    {'fa-sort-asc': !direction},
    {'active': show},
  );
  return (
    <i className={className}></i>
  );
}

const Button = ({ children, ...rest }) => 
  <button { ...rest }>
    {children}
  </button>

const ButtonWithLoading = withLoading(Button);

/**
 * helpers
 */

const buttonStyles = active =>
  classNames('Button inline', {'active': active})

const DateString = ({ DATE }) => {
  return new Date(DATE*1000).toDateString();
}

// heart of the app!
const SOURCES = ['HN', 'Reddit'];

export { Button, withLoading, withError, SortArrow, ButtonWithLoading,
  buttonStyles, SourceIcon, DateString, SOURCES };
