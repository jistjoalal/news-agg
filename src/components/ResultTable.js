import React, { Component } from 'react';
import { sortBy } from 'lodash';
import classNames from 'classnames';

import { Button, SortArrow, withError } from './generic';

const [ lg, md, sm ] = [{width: '40%'}, {width: '30%'}, {width: '10%'}];

const COLUMNS = {
  HN: {
    TITLE: 'title',
    DATE: 'created_at_i',
    COMMENTS: 'num_comments',
    POINTS: 'points',
    URL: 'url',
    ID: 'objectID',
    COMMENTS_URL: 'https://news.ycombinator.com/item?id='
  },
  Reddit: {
    TITLE: 'title',
    DATE: 'created',
    COMMENTS: 'num_comments',
    POINTS: 'score',
    URL: 'url',
    ID: 'id',
    COMMENTS_URL: 'https://news.ycombinator.com/item?id='
  }
};

const SORTS = source => {
  return {
    NONE: list => list,
    TITLE: list => sortBy(list, COLUMNS[source].TITLE),
    DATE: list => sortBy(list, COLUMNS[source].DATE),
    COMMENTS: list => sortBy(list, COLUMNS[source].COMMENTS).reverse(),
    POINTS: list => sortBy(list, COLUMNS[source].POINTS).reverse(),
  }
};

// table formatted for results from HN API
class Table extends Component {
  constructor(props) {
    super(props);
    this.state = {
      sortKey: 'NONE',
      isSortReverse: false,
    };

    this.onSort = this.onSort.bind(this);
  }

  onSort(sortKey) {
    const { isSortReverse } = this.state;
    // toggle if reverse sort if already clicked
    const newISR = this.state.sortKey === sortKey && !isSortReverse;
    this.setState({ sortKey, isSortReverse: newISR });
  }

  render() {
    const { list, onDismiss, source } = this.props;
    const { sortKey, isSortReverse } = this.state;
    const sortedList = SORTS(source)[sortKey](list);
    const toggleSortedList = isSortReverse ? sortedList.reverse() : sortedList;
    return (
      <div className="table">
        <div className="table-header">
          <span style={lg}>
            <Sort sortKey={'TITLE'} onSort={this.onSort}
              activeSortKey={sortKey} isSortReverse={isSortReverse}
            >
              Title
            </Sort>
          </span>
          <span style={md}>
            <Sort sortKey={'DATE'} onSort={this.onSort}
              activeSortKey={sortKey} isSortReverse={isSortReverse}
            >
              Date
            </Sort>
          </span>
          <span style={sm}>
            <Sort sortKey={'COMMENTS'} onSort={this.onSort}
              activeSortKey={sortKey} isSortReverse={isSortReverse}
            >
              Comments
            </Sort>
          </span>
          <span style={sm}>
            <Sort sortKey={'POINTS'} onSort={this.onSort}
              activeSortKey={sortKey} isSortReverse={isSortReverse}
            >
              Points
            </Sort>
          </span>
          <span style={sm}>
            <i className="fa fa-trash"></i>
          </span>
        </div>
        {toggleSortedList.map((item, i) => {
          return (
            <TableRow key={i} source={source}
              item={item} onDismiss={onDismiss}/>
          );
        })}
      </div>
    );
  }
}

const itemBySource = (source, item) => {
  let result = {};
  Object.entries(COLUMNS[source]).forEach(c =>
    result[c[0]] = item[c[1]]
  );
  return result;
}

const DateString = ({ DATE }) => {
  return new Date(DATE*1000).toDateString();
}

const TableRow = ({ item, onDismiss, source }) => {
  const { URL, TITLE, DATE, ID, COMMENTS, COMMENTS_URL,
    POINTS } = itemBySource(source, item);
  return (
    <div key={ID} className="table-row">
      <span style={lg}>
        <a href={URL}>{TITLE}</a>
      </span>
      <span style={md}>
        <DateString DATE={DATE} />
      </span>
      <span style={sm}>
        <a href={`${COMMENTS_URL}${ID}`}>
          {COMMENTS}
        </a>
      </span>
      <span style={sm}>
        {POINTS}
      </span>
      <span style={sm}>
        <Button
          onClick={() => onDismiss(ID)}
          className="button-inline"
        >
          Dismiss
        </Button>
      </span>
    </div>
  );
}
const Sort = ({ sortKey, onSort, children, activeSortKey, isSortReverse }) => {
  const active = sortKey === activeSortKey;
  const direction = isSortReverse ? 'down' : 'up';
  const sortClass = classNames('button-inline', { 'button-active': active });
  return (
    <span>
      <Button onClick={() => onSort(sortKey)} className={sortClass}>
        {children}
      </Button>
      <SortArrow show={active} direction={direction} />
    </span>
  );
}

const ResultTable = withError(Table);

export default ResultTable;