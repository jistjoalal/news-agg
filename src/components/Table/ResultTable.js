import React, { Component } from 'react';
import { sortBy } from 'lodash';

import { withError, COLUMNS } from '../generic';
import HeaderRow from './HeaderRow';
import TableRow from './TableRow';

// table formatted for results from API request
class Table extends Component {
  constructor(props) {
    super(props);
    this.state = {
      sortKey: 'NONE',
      isSortReverse: false,
    };
  }

  // sort table by this key
  onSort = sortKey => {
    this.selectSort(sortKey);
    this.setState({ sortKey });
  }

  // toggle reverse if column already selected
  selectSort = sortKey => {
    if (this.state.sortKey === sortKey) {
      this.setState({ isSortReverse: !this.state.isSortReverse })
    }
  }

  render() {
    const { list, onDismiss, source } = this.props;
    const { sortKey, isSortReverse } = this.state;
    // sort list
    const sortedList = sortBy(list, COLUMNS[source][sortKey.toUpperCase()])
    // toggle reverse
    const toggleSortedList = isSortReverse ? sortedList.reverse() : sortedList;
    return (
      <div className="table">
        <HeaderRow onSort={this.onSort} sortKey={sortKey}
          isSortReverse={isSortReverse} />
        {toggleSortedList.map((item, i) =>
          <TableRow key={i} source={source}
            item={item} onDismiss={onDismiss}/>
        )}
      </div>
    );
  }
}

const ResultTable = withError(Table);
export default ResultTable;