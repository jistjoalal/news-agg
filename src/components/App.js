import React, { Component } from 'react';
import axios from 'axios';

import '../styles/App.scss';
import Search from './Search';
import ResultTable from './ResultTable';
import SourceSelect from './Sources';
import { Button, ButtonWithLoading, withSource, COLUMNS } from './generic';

// TODO: refactor
class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      results: null,
      searchKey: '',
      searchTerm: 'react',
      error: null,
      isLoading: false,
      source: 'HN',
      needFetch: true,
    };
  }
  
  componentDidMount() {
    this.setState(prevState => {
      return { searchKey: prevState.searchTerm }
    });
  }

  componentDidUpdate() {
    const { searchTerm, source } = this.state;
    if (this.state.needFetch) {
      if (this.needsToSearchTopStories(searchTerm, source)) {
        this.fetchStories(searchTerm);
      }
      this.setState({ needFetch: false });
    }
  }
  
  // TODO: "back to top" button
  render() {
    const { searchTerm, results, searchKey, error, isLoading,
      isSortReverse, source }  = this.state;
    const isSaved = this.resultsSaved(results, source, searchKey);
    const list = isSaved ? isSaved.hits : [];
    
    return (
      <div className="page" id="top">
        <div className="interactions">
          <SourceSelect source={source} onChange={this.onSourceChange} />
          <Search
            value={searchTerm}
            onChange={this.onSearchChange}
            onSubmit={this.onSearchSubmit}
            onFocus={this.onFocus}
          >
            Search
          </Search>
          <ResultTable
            list={list}
            onDismiss={this.onDismiss}
            isSortReverse={isSortReverse}
            error={error}
            source={source}
          />
        </div>
        <div className="interactions footer">
          <ButtonWithLoading
            className="button-clickable"
            isLoading={isLoading}
            onClick={() => this.fetchMoreStories(searchKey)}
          >More
          </ButtonWithLoading>
          <a href="#top" className="backtotop">
            <Button className="button-clickable" onClick={() => true}>
                Back to top
            </Button>
          </a>
        </div>
      </div>
    );
  }

  fetchMoreStories = searchTerm => {
    const { source, results } = this.state;
    const isSaved= this.resultsSaved(results, source, searchTerm);
    if (source === 'HN') {
      const page = isSaved ? isSaved.page : 0;
      this.fetchStories(searchTerm, page + 1);
    }
    else if (source === 'Reddit') {
      const after = isSaved ? isSaved.page : 0;
      this.fetchStories(searchTerm, after);
    }
  }

  // makes request to API
  fetchStories = (searchTerm, page = 0) => {
    this.setState({ isLoading: true, error: null });
    axios(withSource(this.state.source)(searchTerm, page))
      .then(result => this.setSearchTopStories(result.data))
      .catch(error => this.setState({ error, isLoading: false }))
  }
  
  // stores response in state.results, appending to existing hits
  setSearchTopStories = response => {
    if (this.state.source === 'HN') {
      const { hits, page } = response;
      this.setState(this.updateSearchTopStories(hits, page));
    }
    else if (this.state.source === 'Reddit') {
      const children = response.data.children.map(c => c.data);
      const after = response.data.after;
      this.setState(this.updateSearchTopStories(children, after));
    }
  }

  onSourceChange = event => {
    this.setState({ source: event.target.value, needFetch: true });
  }

  // clear search input on focus
  onFocus = () => {
    this.setState({searchTerm: ''});
  }
  
  // control search input
  onSearchChange = event => {
    this.setState({ searchTerm: event.target.value });
  }

  // submit search input
  onSearchSubmit = e => {
    const { searchTerm, source } = this.state;
    this.setState(prevState => {
      return { searchKey: prevState.searchTerm }
    });

    if (this.needsToSearchTopStories(searchTerm, source)) {
      this.fetchStories(searchTerm);
    }
    e.preventDefault();
  }

  // remove hit from result list
  onDismiss = id => {
    this.setState(prevState => {
      const { searchKey, results, source } = prevState;
      const { hits, page } = results[source][searchKey];
      const ID_COL = COLUMNS[source].ID;
      const updatedHits = hits.filter(item => item[ID_COL] !== id);
      return {
        results: {
          [source]: {
            ...results,
            [searchKey]: { hits: updatedHits, page }
          }
        }
      };
    });
  }

  // is result for searchTerm cached already?
  needsToSearchTopStories = (searchTerm, source) => {
    const { results } = this.state;
    if (results && results[source] && results[source][searchTerm]) {
      return false;
    }
    return true;
  }

  resultsSaved = (results, source, searchKey) => {
    return results && results[source] && results[source][searchKey];
  }

  // appends additional hits from 'more' button to page
  updateSearchTopStories = (hits, page) => prevState => {
    const { searchKey, results, source } = prevState;
    const oldHits = this.resultsSaved(results, source, searchKey)
      ? results[source][searchKey].hits : [];
    const updatedHits = [...oldHits, ...hits];
    return {
      results: {
        ...results,
        [source]: { 
          [searchKey]: { hits: updatedHits, page }
        }
      },
      isLoading: false
    }
  }
}

export default App;
