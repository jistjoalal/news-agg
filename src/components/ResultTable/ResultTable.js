import React, { Component } from 'react';
import axios from 'axios';

import { unpackResponse, sourceNextPage, withSource,
  COLUMNS } from './helpers';

import Table from './Table';
import Footer from './Footer';

class ResultTable extends Component {
  constructor(props) {
    super(props);
    this.state = {
      results: null,
      isSortReverse: false,
      isLoading: false,
      error: null,
      lastSource: null,
      lastSearchKey: null,
    }
  }

  // initial fetch
  componentDidMount() {
    this.setState({ lastSource: 'fetch pls' });
  }

  // need to fetch results?
  componentDidUpdate() {
    const { lastSource, lastSearchKey, results, isLoading } = this.state;
    const { source, searchKey } = this.props;
    // change of source or searchKey & !already fetching
    if ((lastSource !== source || searchKey !== lastSearchKey) && !isLoading) {
      // results not yet cached
      if (!this.resultsSaved(results, source, searchKey)) {
        this.fetchStories();
      }
      this.setState({ lastSource: source, lastSearchKey: searchKey });
    }
  }

  render() {
    const { results, isLoading, isSortReverse, error } = this.state;
    const { source, searchKey } = this.props;

    // list = saved results or empty list
    const isSaved = this.resultsSaved(results, source, searchKey);
    const list = isSaved ? isSaved.hits : [];
    return (
      <div className="ResultTable">
        <Table list={list} onDismiss={this.onDismiss} onClear={this.dismissAll}
          isSortReverse={isSortReverse} error={error} source={source} />

        <Footer isLoading={isLoading}
          fetchMore={() => this.fetchMoreStories()} />
      </div>
    );
  }

  // results[source][searchKey] cached?
  resultsSaved = () => {
    const { results } = this.state;
    const { source, searchKey } = this.props;
    return results && results[source] && results[source][searchKey];
  }

  // the more button in footer fetches another page of stories
  fetchMoreStories = () => {
    const { results, lastSource } = this.state;
    const { source, searchKey } = this.props;

    // page = current page or 0
    const isSaved = this.resultsSaved(results, source, searchKey);
    const page = isSaved ? isSaved.page : 0;

    // grab another page of stories from source
    this.fetchStories(sourceNextPage(lastSource, page));
  }

  // makes request to API
  // - optional page parameter for fetching next pages of results
  fetchStories = (page = 0) => {
    const { source, searchKey } = this.props;

    // loading icon, clear error
    this.setState({ isLoading: true, error: null });

    // make request
    axios(withSource(source)(searchKey, page))
      .then(res => this.onResponse(res))
      .catch(error => this.setState({ error, isLoading: false }))
  }
  
  // interprets response according to source API
  onResponse = ({ data }) => {
    this.cacheResults(unpackResponse(this.state.lastSource, data));
  }

  // caches fetched results
  cacheResults = ({ hits, page }) => {
    this.setState(({ results, lastSource, lastSearchKey }) => {
    // catch null results (cant spread null):
    //   null results[source] = {}
    //   null results[source][searchKey] = {}
    //   null results[source][searchKey].hits = []
    const sourceResults = results && results[lastSource] ?
      {...results}[lastSource] : {};
    const searchResults = sourceResults && sourceResults[lastSearchKey] ?
      {...sourceResults}[lastSearchKey] : {};
    const oldHits = searchResults.hits || [];
    return {
      // keep old results (other sources)
      results: { ...results,
        // keep old results[source] (other searchKeys)
        [lastSource]: { ...sourceResults,
          [lastSearchKey]: {
            // append new hits + page
            hits: [...oldHits, ...hits],
            page
          }
        }
      },
      isLoading: false
    }
  })}

  dismissAll = () => this.setState(({ results, lastSource, lastSearchKey }) => {
    return {
      results: { ...results,
        [lastSource]: {...results[lastSource],
          [lastSearchKey]: null
        }
      }
    }
  })

  // remove hit from result list
  onDismiss = id => this.setState(({ results, lastSource, lastSearchKey }) => {
    const { hits, page } = results[lastSource][lastSearchKey];

    // filter out this table row from results
    const updatedHits = hits.filter(item =>
      item[COLUMNS[lastSource].ID] !== id);
    return {
      results: { ...results,
        [lastSource]: { ...results[lastSource],
          [lastSearchKey]: { hits: updatedHits, page }
        }
      }
    };
  })
}

export default ResultTable;