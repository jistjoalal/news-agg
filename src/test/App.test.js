import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });

import App from '../components/App';
import ResultTable from '../components/ResultTable';
import Search from '../components/Search';
import { Button } from '../components/generic';

describe('App', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('has a valid snapshot', () => {
    const component = renderer.create(<App />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Search', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Search>Search</Search>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('has a valid snapshot', () => {
    const component = renderer.create(<Search>Search</Search>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Button', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Button>Button</Button>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('has a valid snapshot', () => {
    const component = renderer.create(<Button>Button</Button>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('displays the button text', () => {
    const element = shallow(<Button>hello</Button>);
    expect(element.props().children).toBe('hello');
  })
});

describe('ResultTable', () => {
  const props = {
    list: [
      { title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y' },
      { title: '2', author: '2', num_comments: 1, points: 2, objectID: 'x' }
    ],
    sortKey: 'TITLE',
    isSortReverse: false,
    source: 'HN',
  };

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<ResultTable { ...props } />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('has a valid snapshot', () => {
    const component = renderer.create(<ResultTable { ...props } />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});