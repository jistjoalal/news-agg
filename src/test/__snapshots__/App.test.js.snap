// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <div
      className="SourceSelect"
    >
      <input
        checked={true}
        className="SourceSelect-radio"
        onChange={[Function]}
        type="radio"
        value="HN"
      />
      HN
      <input
        checked={false}
        className="SourceSelect-radio"
        onChange={[Function]}
        type="radio"
        value="Reddit"
      />
      Reddit
    </div>
    <form
      className="search"
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        onFocus={[Function]}
        type="text"
        value="react"
      />
      <button
        className="button-clickable"
        type="submit"
      >
        Search
      </button>
    </form>
    <div
      className="table"
    >
      <div
        className="table-header"
      >
        <span
          style={
            Object {
              "width": "40%",
            }
          }
        >
          <span>
            <button
              className="button-inline"
              onClick={[Function]}
            >
              Title
            </button>
          </span>
        </span>
        <span
          style={
            Object {
              "width": "30%",
            }
          }
        >
          <span>
            <button
              className="button-inline"
              onClick={[Function]}
            >
              Date
            </button>
          </span>
        </span>
        <span
          style={
            Object {
              "width": "10%",
            }
          }
        >
          <span>
            <button
              className="button-inline"
              onClick={[Function]}
            >
              Comments
            </button>
          </span>
        </span>
        <span
          style={
            Object {
              "width": "10%",
            }
          }
        >
          <span>
            <button
              className="button-inline"
              onClick={[Function]}
            >
              Points
            </button>
          </span>
        </span>
        <span
          style={
            Object {
              "width": "10%",
            }
          }
        >
          <i
            className="fa fa-trash"
          />
        </span>
      </div>
    </div>
  </div>
  <div
    className="interactions"
  >
    <i
      className="fa fa-spinner fa-spin fa-3x"
    />
  </div>
</div>
`;

exports[`Button has a valid snapshot 1`] = `
<button
  className=""
>
  Button
</button>
`;

exports[`ResultTable has a valid snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-header"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <span>
        <button
          className="button-inline"
          onClick={[Function]}
        >
          Title
        </button>
      </span>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <span>
        <button
          className="button-inline"
          onClick={[Function]}
        >
          Date
        </button>
      </span>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <span>
        <button
          className="button-inline"
          onClick={[Function]}
        >
          Comments
        </button>
      </span>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <span>
        <button
          className="button-inline"
          onClick={[Function]}
        >
          Points
        </button>
      </span>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <i
        className="fa fa-trash"
      />
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a>
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      Invalid Date
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <a
        href="undefinedy"
      >
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a>
        2
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      Invalid Date
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <a
        href="undefinedx"
      >
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;

exports[`Search has a valid snapshot 1`] = `
<form
  className="search"
>
  <input
    type="text"
  />
  <button
    className="button-clickable"
    type="submit"
  >
    Search
  </button>
</form>
`;
